Fix `Parallax2D` repeats being not relative to its transform

kleonc opened the pull request:
Fixes #95574.

Fixes the regression regarding the behavior of ParallaxBackground/ParallaxLayer.
Makes the Parallax2D behave in the same manner.

Project with examples presented below (Parallax2D only): [parallax_demo_MRP.zip](https://github.com/user-attachments/files/16642385/parallax_demo_MRP.zip)
Original|repeat setup|Before<br>(v4.3.stable.official [77dcf97d8])|After<br>(this PR)
-|-|-|-
![ekp9pTmq3l](https://github.com/user-attachments/assets/76231330-13b7-4484-98de-2793c3ae4b0d)|![QD4bNZda4O](https://github.com/user-attachments/assets/92e837fd-90e2-477a-a12b-2e1be551ae51)|![JufFXTox5u](https://github.com/user-attachments/assets/47d5c261-ddf0-40f3-8800-e236a561660d)|![godot windows editor dev x86_64_u2YnsrKQ4F](https://github.com/user-attachments/assets/239ff5d0-f332-4518-b9ec-edfc0895df10)
![zxxzYg9bp8](https://github.com/user-attachments/assets/edc2b77b-ed84-4b0b-ad3d-b172af9e7c57)|![Godot_v4 3-stable_win64_nWGuRWpn78](https://github.com/user-attachments/assets/8a084758-89bf-4dc8-84d3-e25a1e73be89)|![1t9cwc8SM2](https://github.com/user-attachments/assets/3b211de7-a660-4117-a32a-1968489f316a)|![YlLphNqNc4](https://github.com/user-attachments/assets/a26dc098-c16b-47d4-a04a-1cfdb961f808)

kleonc commented on drivers/gles3/rasterizer_canvas_gles3.cpp:
```diff
@@ -647,18 +647,17 @@ void RasterizerCanvasGLES3::_render_items(RID p_to_render_target, int p_item_cou
-				pos.x += ci->repeat_size.x;
-				pos.y = start_pos.y;
-			} while (pos.x < end_pos.x);
+			Point2 offset;
+
+			int repeat_times_x = ci->repeat_size.x ? ci->repeat_times : 0;
+			int repeat_times_y = ci->repeat_size.y ? ci->repeat_times : 0;
+			for (int ry = 0; ry <= repeat_times_y; ry++) {
+				offset.y = start_pos.y + ry * ci->repeat_size.y;
+				for (int rx = 0; rx <= repeat_times_x; rx++) {
+					offset.x = start_pos.x + rx * ci->repeat_size.x;
+					_record_item_commands(ci, p_to_render_target, p_canvas_transform_inverse, current_clip, blend_mode, p_lights, index, batch_broken, r_sdf_used, offset);
+				}
+			}
```
This is not directly related but changed it to use multiplication over addition, as it can potentially result in accumulating floating points errors and thus one more/less repeat per axis than desired.

Rather unlikely but better safe than sorry. Previously failing example (v4.3.stable.official [77dcf97d8]):
![Godot_v4 3-stable_win64_Lvpgl7eGEb](https://github.com/user-attachments/assets/9371db47-b604-48f5-a1d7-1f6c973e9cd0)

Also changed it to more common row-by-row iteration order.

kleonc commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -435,10 +442,14 @@ void RendererCanvasCull::canvas_set_item_mirroring(RID p_canvas, RID p_item, con
 }
 
 void RendererCanvasCull::canvas_set_item_repeat(RID p_item, const Point2 &p_repeat_size, int p_repeat_times) {
+	ERR_FAIL_COND(p_repeat_size.x < 0 || p_repeat_size.y < 0);
+	ERR_FAIL_COND(p_repeat_times < 0);
```
While disallowing negative `p_repeat_times` is kinda obvious, I'm not sure about `p_repeat_size`. Should I remove the first check? :thinking:

markdibarry replied:
Would it even work with a negative repeat size? I don't think its set up to work backwards, but it's been awhile since I looked.

kleonc replied:
Parallax2D prevents that, but manually using RenderingServer it was possible to set negative repeat size. I think it might somehow work, but I'd suspect rect/clipping calculations doesn't account for that so likely not working correctly anyway.

kleonc replied:
I've end up removing that `ERR_FAIL_COND(p_repeat_size.x < 0 || p_repeat_size.y < 0);` check (and not adding similar check in `RendererCanvasCull::canvas_set_item_mirroring`).

With the current code it should work fine for a negative repeat size, would simply result in adding repeats in the opposite direction first (rect would still be calculated correctly).

So e.g. for `p_repeat_times = 1`:
- `p_repeat_size = Vector2(64, 0)` would add a single repeat to the right (+X),
- `p_repeat_size = Vector2(-64, 0)` would add a single repeat to the left (-X),
- `p_repeat_size = Vector2(0, 64)` would add a single repeat to the bottom (+Y),
- `p_repeat_size = Vector2(0, -64)` would add a single repeat to the top (-Y).

For even `p_repeat_times` results should be the same regardless of the size's signs (same repeat count per right/left and top/bottom).

Note that both Parallax2D/ParallaxLayer do prevent negative repeat sizes themselves, nothing changes about it. But since negative size works in a sensible manner for the current implementation I see no point in preventing such usage when manually using the RenderingServer.

markdibarry commented:
This is great! I was working on this right now, and I was thinking of similar changes as well. It seems you have the same issue I'm tackling too, the culling still is expecting the transform to follow the original, so it breaks:
![image](https://github.com/user-attachments/assets/9dfce7ba-d267-42b0-9af5-b5722109be16)

That being said, I agree with the change to the do-while mainly because I just hated having that there at all. Second, if we're passing the source canvas item, might as well remove the repeat size and times params since we can access them directly on the source.

Thanks for digging in with me on some nagging issues!

kleonc commented:
> It seems you have the same issue I'm tackling too, the culling still is expecting the transform to follow the original, so it breaks:

Indeed I haven't looked into rects/clipping (yet!). :upside_down_face: Will do so tomorrow and update this.

> Second, if we're passing the source canvas item, might as well remove the repeat size and times params since we can access them directly on the source.

Duh! :man_facepalming:

markdibarry commented:
This also would work in the renderer and would prevent us from needing to add even more bloat to `CanvasItem`, but I'm not sure if there's a better way or how bad this is performance-wise compared to passing the source item everywhere

```cpp
Transform2D base_transform;

if (p_offset.x || p_offset.y) {
	Transform2D new_transform = p_item->xform_curr;
	new_transform.set_origin(new_transform.get_origin() + p_offset);
	new_transform = p_item->final_transform * p_item->xform_curr.affine_inverse() * new_transform;
	base_transform = p_canvas_transform_inverse * new_transform; // TODO: Interpolate or explain why not needed.
} else {
	base_transform = p_canvas_transform_inverse * p_item->final_transform;
}
```

kleonc commented:
> This also would work in the renderer and would prevent us from needing to add even more bloat to `CanvasItem`, but I'm not sure if there's a better way or how bad this is performance-wise compared to passing the source item everywhere
> 
> ```c++
> Transform2D base_transform;
> 
> if (p_offset.x || p_offset.y) {
> 	Transform2D new_transform = p_item->xform_curr;
> 	new_transform.set_origin(new_transform.get_origin() + p_offset);
> 	new_transform = p_item->final_transform * p_item->xform_curr.affine_inverse() * new_transform;
> 	base_transform = p_canvas_transform_inverse * new_transform; // TODO: Interpolate or explain why not needed.
> } else {
> 	base_transform = p_canvas_transform_inverse * p_item->final_transform;
> }
> ```

This would _kinda_ work for direct children of the repeat source (Parallax2D), but not in general case. The offset needs to be applied relative to the repeat source (Parallax2D node)(hence why I concluded simplest would be to store a reference to it). Try the animated scene example from the project I've attached, won't work correctly.

Moreover `xform_curr` is potentially different than the actual current transform because of physics interpolation. Aka using it for calculations would introduce additional discrepancies (thus "would _kinda_ work").

`final_transform` which I've used for calculations already incorporate physics interpolation, so there shouldn't be issues with it (haven't actually tested it with physics interpolation).

Also for reference my current calculation:
```cpp
Transform2D base_transform = p_item->final_transform;
if (p_item->repeat_source_item && (p_repeat_offset.x || p_repeat_offset.y)) {
	base_transform.columns[2] += p_item->repeat_source_item->final_transform.basis_xform(p_repeat_offset);
}
base_transform = p_canvas_transform_inverse * base_transform;
```
is equivalent to / a simplified version of:
```cpp
Transform2D base_transform = p_item->final_transform;
if (p_item->repeat_source_item && (p_repeat_offset.x || p_repeat_offset.y)) {
	Transform2D source_xform = p_item->repeat_source->final_transform;
	base_transform = source_xform * Transform2D(0, p_repeat_offset) * source_xform.affine_inverse() * base_transform;
}
base_transform = p_canvas_transform_inverse * base_transform;
```

kleonc commented:
Updated the rect calculation, further testing welcomed. :slightly_smiling_face: If I got it right it should result in the white rect (contrary to the blue one which is seemingly simpler to obtain):
![YMOdjvjAOn](https://github.com/user-attachments/assets/0258715a-217c-4a7d-8d00-332a8598cf47)

markdibarry commented:
Doing some general performance tests before/after.

I agree that it's better to go add more stuff and do it the "right way" rather than hacking it with the interpolation data. It's beneficial to have the repeat_source_item anyway, just because that opens us up to being able to support nested `Parallax2D` eventually. I just always take pause when I'm throwing more logic and stuff onto objects in a hot-path.

markdibarry commented:
Also looks like the culling is correct. CanvasGroup is so good for visualizing it:
 

https://github.com/user-attachments/assets/444b6854-1711-4408-9507-6466b142abd4

markdibarry commented:
Hm. Unfortunately, this seems to break `ParallaxBackground` and `ParallaxLayer`s, which seem to not repeat anymore. We'll have to make sure there's no regressions there.

markdibarry commented:
@kleonc As one solution, we could just have `ParallaxLayer` store data the same as `Parallax2D`. I don't see any side effects from any of my tests with the following changes:

`parallax_layer.cpp line 65-78`
```cpp
void ParallaxLayer::_update_mirroring() {
	if (!is_inside_tree()) {
		return;
	}

	ParallaxBackground *pb = Object::cast_to<ParallaxBackground>(get_parent());
	if (pb) {
		RID c = pb->get_canvas();
		RID ci = get_canvas_item();
		Point2 mirror_scale = mirroring * orig_scale;
		RenderingServer::get_singleton()->canvas_set_item_repeat(ci, mirror_scale, 1);
		RenderingServer::get_singleton()->canvas_item_set_interpolated(ci, false);
	}
}
```

`renderer_canvas_cull.cpp line 53-55`
```cpp
for (int i = 0; i < p_child_item_count; i++) {
	_cull_canvas_item(p_child_items[i].item, p_transform, p_clip_rect, Color(1, 1, 1, 1), 0, z_list, z_last_list, nullptr, nullptr, true, p_canvas_cull_mask, Point2(), 1, nullptr);
}
```

We'd also be free to remove `canvas_set_item_mirroring()` from the rendering server in that case as well.

markdibarry commented:
Also, did some testing with 40 `Parallax2D`s, with 5 repeats each (You shouldn't need it higher than 3, though I know people will use 20 if they can) I was getting between 7500-7000 FPS on my machine and no noticeable change between branches.

Since taking advantage of the changes to the parallax logic, there's similar performance for `ParallaxBackground` and `ParallaxLayer` as well. Just out of curiousity, I did a similar test for 40 `ParallaxLayer`s on 4.2 (obviously without the 5 repeats) and it was somewhere around 1800 FPS, so... wow!

kleonc commented:
> Hm. Unfortunately, this seems to break `ParallaxBackground` and `ParallaxLayer`s, which seem to not repeat anymore. We'll have to make sure there's no regressions there.

I guess I forgot they exist. :laughing: But to be clear you're saying they should repeat the same way as Parallax2D with this PR, not in the pre-this-PR manner?




> @kleonc As one solution, we could just have `ParallaxLayer` store data the same as `Parallax2D`. I don't see any side effects from any of my tests with the following changes:
> 
> `parallax_layer.cpp line 65-78`
> 
> ```c++
> void ParallaxLayer::_update_mirroring() {
> 	if (!is_inside_tree()) {
> 		return;
> 	}
> 
> 	ParallaxBackground *pb = Object::cast_to<ParallaxBackground>(get_parent());
> 	if (pb) {
> 		RID c = pb->get_canvas();
> 		RID ci = get_canvas_item();
> 		Point2 mirror_scale = mirroring * orig_scale;
> 		RenderingServer::get_singleton()->canvas_set_item_repeat(ci, mirror_scale, 1);
> 		RenderingServer::get_singleton()->canvas_item_set_interpolated(ci, false);
> 	}
> }
> ```
> 
> `renderer_canvas_cull.cpp line 53-55`
> 
> ```c++
> for (int i = 0; i < p_child_item_count; i++) {
> 	_cull_canvas_item(p_child_items[i].item, p_transform, p_clip_rect, Color(1, 1, 1, 1), 0, z_list, z_last_list, nullptr, nullptr, true, p_canvas_cull_mask, Point2(), 1, nullptr);
> }
> ```

Currently (pre-this-PR) `mirror` set in `RenderingServer.canvas_set_item_mirroring` is used only as the initial value for `_cull_canvas_item`'s `p_repeat_size` argument:
https://github.com/godotengine/godot/blob/1bd740d18d714f815486b04bf4c6154ef6c355d9/servers/rendering/renderer_canvas_cull.cpp#L54

Meaning if a canvas item is set as repeat source (e.g. via directly calling `RenderingServer.canvas_set_item_repeat`) then it would already ignore such mirroring. So there's a potential clash between repeat/mirroring already, thus I agree it makes sense to make both use repeat within the RenderingServer (seems they are already not meant to be used together).

But instead of changing `ParallaxLayer` to call `RenderingServer.canvas_set_item_repeat` I think it makes more sense to change `RenderingServer.canvas_set_item_mirroring` to do the same as `RenderingServer.canvas_set_item_repeat` (or just call it). And remove the `mirror` from the RenderingServer internals as it would be no longer used.


> We'd also be free to remove `canvas_set_item_mirroring()` from the rendering server in that case as well.

While I agree that `RenderingServer.canvas_set_item_mirroring` seems absolutely useless, I don't think we can just remove it, would be compatibility breaking. But I think deprecating it should be fine, and the docs could be updated to recommend using `RenderingServer.canvas_set_item_repeat` instead (as it would be a wrapper for it anyway).

> Also, did some testing with 40 `Parallax2D`s, with 5 repeats each (You shouldn't need it higher than 3, though I know people will use 20 if they can) I was getting between 7500-7000 FPS on my machine and no noticeable change between branches.

Kinda expected no noticeable change as not doing anything heavy in here, but great to see a confirmation!

> Since taking advantage of the changes to the parallax logic, there's similar performance for `ParallaxBackground` and `ParallaxLayer` as well. Just out of curiousity, I did a similar test for 40 `ParallaxLayer`s on 4.2 (obviously without the 5 repeats) and it was somewhere around 1800 FPS, so... wow!

Comparing with the numbers from https://github.com/godotengine/godot/pull/87391#issue-2091849239 seems like it got better in the meantime, great to see!

@markdibarry Massive thanks for testing/feedback! :slightly_smiling_face:

markdibarry commented:
> I guess I forgot they exist. 😆 But to be clear you're saying they should repeat the same way as Parallax2D with this PR, not in the pre-this-PR manner?

This is a regression in `ParallaxLayer` and an oversight in `Parallax2D`. In 4.2 `ParallaxLayer` behaved like `Parallax2D` does after this PR, so it should definitely get the behavior back.

> But instead of changing ParallaxLayer to call RenderingServer.canvas_set_item_repeat I think it makes more sense to change RenderingServer.canvas_set_item_mirroring to do the same as RenderingServer.canvas_set_item_repeat (or just call it). And remove the mirror from the RenderingServer internals as it would be no longer used.

That's a good point. I think the plan is to deprecate on the next version, and promote Parallax2D to non-experimental, so that does make more sense. Lets go with that!

> Comparing with the numbers from https://github.com/godotengine/godot/pull/87391#issue-2091849239 seems like it got better in the meantime, great to see!

Well, it's semi-arbitrary, since an empty scene is 8000 FPS for my machine now, and that could be because I've got a new compy, but also could be multiple other variables. Either way, seeing that there's more than 4x improvement between versions for heavier scenes even for the old system is very promising.

markdibarry commented on servers/rendering/renderer_canvas_cull.cpp:
```diff
@@ -291,6 +292,27 @@ void RendererCanvasCull::_cull_canvas_item(Item *p_canvas_item, const Transform2
+		Rect2 corner_rect = global_rect;
+		corner_rect.position -= (repeat_times / 2) * (repeat_size_x_offset + repeat_size_y_offset);
+		global_rect = corner_rect;
+
```
You need to multiply the `repeat_times` to the offsets for the other corners.

kleonc replied:
Oh, right, simplified it too much while refactoring a little. :upside_down_face: Nice catch!

Actually, instead of multiplying by `repeat_times` here I'll revert back to the previous version with 3 `basis_xform()`s (instead of 2 like here), as multiplying the results would be a little more floating point error prone (and `basis_xform` is cheap). It again likely doesn't matter though. :slightly_smiling_face:

markdibarry commented:
If you fix the `repeat_times` issue, I can't find any other problems. LGTM!

kleonc commented:
> This is a regression in `ParallaxLayer` and an oversight in `Parallax2D`. In 4.2 `ParallaxLayer` behaved like `Parallax2D` does after this PR, so it should definitely get the behavior back.

Oh, indeed that's a regression! Another good catch. :+1: So definitely this PR should go into 4.3.x (was going to mark it for cherry-picking anyway).



> Yeah, that's why I replaced it with a default value in my suggestion. I originally passed in the mirroring so the old nodes could use the new system, but adding a source item parameter complicates things. So whether copying the new repeat method logic to the old
> mirror method or calling the new one, it should have the same result. I'll trust your decision.

Same here, I also tried passing the old `mirror`, and concluded that making it use repeating instead was more sensible than patching things along the way. Above I've added some justification for why I think it's fine to do so (and almost for sure shouldn't break anything) only for future reference.

> That's a good point. I think the plan is to deprecate on the next version, and promote Parallax2D to non-experimental, so that does make more sense. Lets go with that!

For sure Parallax2D would need to be marked as non-experimental before deprecating ParallaxLayer/ParallaxBackground (or at the same time). When exactly it's planned to be done I don't know, but makes sense to me to already deprecate `RenderingServer.canvas_set_item_mirroring`. Will probably open a separate PR for that later (not meant to be cherry-picked into 4.3.x).

kleonc committed:
Fix Parallax2D repeats being not relative to its transform

Mickeon commented on doc/classes/RenderingServer.xml:
```diff
@@ -944,7 +944,8 @@
 			<description>
-				A copy of the canvas item will be drawn with a local offset of the mirroring [Vector2].
+				A copy of the canvas item will be drawn with a local offset of the [param mirroring].
+				[b]Note:[/b] Equivalent to calling [method canvas_set_item_repeat] like [code]canvas_set_item_repeat(item, mirroring, 1)[/code], with an additional check ensuring [param canvas] is a parent of [param item].
```
Both `item_mirroring` and `item_repeat` have very... odd descriptions which feel like don't describe what the method actually does. But at least this nitpick could be tweaked for now.
```suggestion
				[b]Note:[/b] This is equivalent to calling [method canvas_set_item_repeat] like [code]canvas_set_item_repeat(item, mirroring, 1)[/code], with an additional check ensuring [param canvas] is a parent of [param item].
```

Mickeon reviewed (approved):
If @markdibarry approves I approve.

akien-mga commented:
Thanks!

akien-mga commented:
Cherry-picked for 4.3.1.
